tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,652 @@
+	tb.beaconsMu.RUnlock()
+
+	if !ok {
+		return types.Hash32{}, ErrBeaconNotCalculated
+	}
+
+	if tb.tortoiseBeaconDB != nil {
+		if err := tb.tortoiseBeaconDB.SetTortoiseBeacon(epochID, beacon); err != nil {
+			return types.Hash32{}, err
+		}
+	}
+
+	return beacon, nil
+}
+
+// Wait waits until beacon for this epoch becomes ready.
+func (tb *TortoiseBeacon) Wait(epochID types.EpochID) error {
+	if tb.tortoiseBeaconDB != nil {
+		if _, ok := tb.tortoiseBeaconDB.GetTortoiseBeacon(epochID); ok {
+			return nil
+		}
+	}
+
+	tb.beaconsMu.RLock()
+	ch, ok := tb.beaconsReady[epochID]
+	tb.beaconsMu.RUnlock()
+
+	if !ok {
+		return ErrBeaconNotCalculated
+	}
+
+	<-ch
+	return nil
+}
+
+func (tb *TortoiseBeacon) cleanupLoop() {
+	ticker := time.NewTicker(cleanupInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-tb.CloseChannel():
+			return
+		case <-ticker.C:
+			tb.cleanup()
+		}
+	}
+}
+
+func (tb *TortoiseBeacon) cleanup() {
+	// TODO(nkryuchkov): implement a better solution, consider https://github.com/golang/go/issues/20135
+	for e := range tb.beacons {
+		if tb.epochIsOutdated(e) {
+			delete(tb.beacons, e)
+			delete(tb.beaconsReady, e)
+		}
+	}
+}
+
+func (tb *TortoiseBeacon) epochIsOutdated(epoch types.EpochID) bool {
+	tb.layerMu.Lock()
+	lastEpoch := tb.lastLayer.GetEpoch()
+	tb.layerMu.Unlock()
+
+	return lastEpoch-epoch > cleanupEpochs
+}
+
+func (tb *TortoiseBeacon) listenMessages() error {
+	for {
+		select {
+		case <-tb.CloseChannel():
+			return nil
+		default:
+			m := tb.messageReceiver.Receive()
+
+			if err := tb.handleMessage(m); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// listens to new layers.
+func (tb *TortoiseBeacon) listenLayers() {
+	for {
+		select {
+		case <-tb.CloseChannel():
+			return
+		case layer := <-tb.layerTicker:
+			tb.Log.Info("Received tick for layer %v", layer)
+			go tb.handleLayer(layer)
+		}
+	}
+}
+
+// the logic that happens when a new layer arrives.
+// this function triggers the start of new CPs.
+func (tb *TortoiseBeacon) handleLayer(layer types.LayerID) {
+	tb.layerMu.Lock()
+	if layer > tb.lastLayer {
+		tb.lastLayer = layer
+	}
+
+	tb.layerMu.Unlock()
+
+	epoch := layer.GetEpoch()
+	tb.Debug("tortoise beacon got tick at layer %v epoch %v", layer, epoch)
+
+	tb.handleEpoch(epoch)
+}
+
+func (tb *TortoiseBeacon) handleEpoch(epoch types.EpochID) {
+	if epoch.IsGenesis() {
+		tb.With().Info("not starting tortoise beacon since we are in genesis epoch", epoch)
+
+		return
+	}
+
+	tb.Log.Info("Calculating beacon for epoch %v", epoch)
+
+	tb.beaconsMu.Lock()
+
+	if _, ok := tb.beacons[epoch]; ok {
+		tb.beaconsMu.Unlock()
+
+		// Already handling this epoch
+		return
+	}
+
+	tb.beaconsReady[epoch] = make(chan struct{})
+	tb.beaconsMu.Unlock()
+
+	// round 0
+	// take all ATXs received in last epoch (i -1)
+	atxList := tb.epochATXGetter.GetEpochAtxs(epoch - 1)
+
+	// concat them into a single proposal message
+	m := NewProposalMessage(epoch, atxList)
+	tb.messageSender.Send(m)
```

Comment:
since this message is sent through gossip - wouldn't the sender receive it as well?